dnf-behave-tests/dnf/plugins-core/pre-transaction-actions.feature

```diff
@@ -0,0 +1,124 @@
+   When I execute dnf with args "install glibc"
+   Then the exit code is 0
+    And Transaction is following
+       | Action        | Package                                   |
+       | install       | glibc-0:2.28-9.fc29.x86_64                |
+       | install-dep   | setup-0:2.12.1-1.fc29.noarch              |
+       | install-dep   | glibc-all-langpacks-0:2.28-9.fc29.x86_64  |
+       | install-dep   | glibc-common-0:2.28-9.fc29.x86_64         |
+       | install-dep   | filesystem-0:3.9-2.fc29.x86_64            |
+       | install-dep   | basesystem-0:11-6.fc29.noarch             |
+    And file "/trans.log" contents is
+    """
+    install glibc-0:2.28-9.fc29.x86_64 repo dnf-ci-fedora
+    """
+
+Examples:
+    | filter            |
+    | /etc/ld.so.conf   |
+    | /etc/ld*conf      |
+    | glibc             |
+    | g*c               |
+
+
+Scenario Outline: I can filter on transaction state
+  Given I create and substitute file "/etc/dnf/plugins/pre-transaction-actions.d/test.action" with
+    """
+    *:<state>:echo '${{state}} ${{name}}' >> {context.dnf.installroot}/trans.log
+    """
+    And I create file "/trans.log" with
+    """
+    """
+   When I execute dnf with args "install setup"
+   Then the exit code is 0
+    And file "/trans.log" contents is
+    """
+    <output>
+    """
+
+Examples:
+    | state     | output            |
+    | any       | install setup     |
+    | in        | install setup     |
+    | out       |                   |
+
+
+Scenario: Do not traceback when reason change is in transaction
+  Given I create and substitute file "/etc/dnf/plugins/pre-transaction-actions.d/test.action" with
+    """
+    *:any:echo '${{state}} ${{name}}-${{epoch}}:${{ver}}-${{rel}}.${{arch}} repo ${{repoid}}' > {context.dnf.installroot}/trans.log
+    """
+    And I use repository "installonly"
+    And I configure dnf with
+        | key                          | value         |
+        | installonlypkgs              | installonlyA  |
+        | installonly_limit            | 2             |
+   When I execute dnf with args "-v install installonlyA-1.0"
+   Then the exit code is 0
+    And Transaction is following
+        | Action        | Package                         |
+        | install       | installonlyA-1.0-1.x86_64       |
+    And stderr does not contain "Traceback"
+    And file "/trans.log" contents is
+    """
+    install installonlyA-0:1.0-1.x86_64 repo installonly
+    """
+   When I execute dnf with args "-v install installonlyA-2.0"
+   Then the exit code is 0
+    And Transaction is following
+        | Action        | Package                         |
+        | install       | installonlyA-2.0-1.x86_64       |
+    And stderr does not contain "Traceback"
+    And file "/trans.log" contents is
+    """
+    install installonlyA-0:2.0-1.x86_64 repo installonly
+    """
+   When I execute dnf with args "-v install installonlyA-2.2"
+   Then the exit code is 0
+    And Transaction is following
```

Comment:
Reason change happen when install and remove kernel in one transaction. This is a workaround that fix logic of history DB in background of DNF.

Reply:
It looks like that reason change is added only in DNF5, and not in DNF4